### ISSUE TITLE
Include profiling statistics in query responses

### ISSUE THREAD

**Issue by domodwyer** [influxdata/influxdb_iox#6055]
A query can source data from a combination of three different places:

* Querier Cache
* Ingester's "hot data"
* S3

Depending on where the data comes from, the latency can vary in a non-trivial way (especially for "quick" queries where data sourcing dominates the query time). This can be confusing for end users who are trying to understand why a query at time T1 was fast, but T2 was slow.

This latency variation is further complicated when you consider a prod deployment with multiple queriers (hitting the same querier twice), cache evictions and restarts/deployments effectively clearing the cache of both the queriers and ingesters.

This information can be gleaned from distributed tracing, but it's tricky to decode and makes retroactive analysis impossible if the user didn't force a trace / it wasn't sampled / the user is running in an environment without a reliable tracing infra setup.

## Proposal

Return "execution" timing traces (including data sources) to the client, perhaps via gRPC trailers.

For example, flux returns profiling data (when requested) that includes timing of various stages of execution, and a snazzy graphviz graph of the process:

![image](https://user-images.githubusercontent.com/9275968/199943944-0a46e4e4-2f62-459c-b087-f7d1ed076611.png)

```
#group,false,false,true,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false
#datatype,string,long,string,long,long,long,long,long,long,long,long,long,string,long,long,string,string,string,long
#default,_profiler,,,,,,,,,,,,,,,,,,
,result,table,_measurement,TotalDuration,CompileDuration,QueueDuration,PlanDuration,RequeueDuration,ExecuteDuration,Concurrency,MaxAllocated,TotalAllocated,RuntimeErrors,influxdb/scanned-values,influxdb/network-read-time,tracing/id,tracing/sampled,flux/query-plan,influxdb/scanned-bytes
,,0,profiler/query,11391590,197039,25273,0,0,11150568,0,40896,51776,,600,16011247,"0620a09df867b5cd
","false
","digraph {
  ""merged_ReadRange5_filter2""
  ""sortedPivot6""
  ""group4""

  ""merged_ReadRange5_filter2"" -> ""sortedPivot6""
  ""sortedPivot6"" -> ""group4""
}

",41491

#group,false,false,true,false,false,false,false,false,false,false
#datatype,string,long,string,string,string,long,long,long,long,double
#default,_profiler,,,,,,,,,
,result,table,_measurement,Type,Label,Count,MinDuration,MaxDuration,DurationSum,MeanDuration
,,1,profiler/operator,*influxdb.readFilterSource,merged_ReadRange5_filter2,1,3219892,3219892,3219892,3219892
,,1,profiler/operator,*universe.sortedPivotTransformation,sortedPivot6,12,930,112227,299504,24958.666666666668
,,1,profiler/operator,*universe.groupTransformationAdapter,group4,4,666,10900,15188,3797
```

It does this via adding trailers to the response.

Loki does something similar when `--stats` is passed on the CLI when executing a query:

```console
$ logcli query <...snip...> --stats
Ingester.TotalReached            96
Ingester.TotalChunksMatched      15
Ingester.TotalBatches            6
Ingester.TotalLinesSent          180
Ingester.TotalChunksRef          0
Ingester.TotalChunksDownloaded  0
Ingester.ChunksDownloadTime      0s
Ingester.HeadChunkBytes          637 kB
Ingester.HeadChunkLines          2903
Ingester.DecompressedBytes       2.9 MB
Ingester.DecompressedLines       12140
Ingester.CompressedBytes         1.4 MB
Ingester.TotalDuplicates         0
Querier.TotalChunksRef   3
Querier.TotalChunksDownloaded    3
Querier.ChunksDownloadTime       14.362802ms
Querier.HeadChunkBytes   0 B
Querier.HeadChunkLines   0
Querier.DecompressedBytes        825 kB
Querier.DecompressedLines        3496
Querier.CompressedBytes          300 kB
Querier.TotalDuplicates          90
Summary.BytesProcessedPerSecond          172 MB
Summary.LinesProcessedPerSecond          737980
Summary.TotalBytesProcessed              4.3 MB
Summary.TotalLinesProcessed              18539
Summary.ExecTime                         25.121265ms
Summary.QueueTime                        0s
```

We should include:

* Where the various bits of data came from
  * How many ingesters were hit, how long that took, etc
* How long fetching each bit of data took
* If it was a cache hit/miss
* Which querier executed it (so that we can identify sticky-routing failures/misses)
* How much data this churned through
  * Bytes downloaded from the object store
  * DF execution timing
  * DF execution stats?
* Probably more! 

Internal Influx teams make use of this sort of timing data to diagnose and debug issues, both internal and for customers.

**Comment by alamb** [comment]
There is a non trivial amount of profiling available now (that currently can be seen in the distributed traces and see in jaeger). Adding some/all of this information to the grpc trailer would be likely be straightforward. 

In particular, here is the code that converts the Datafusion metrics to tracing spans: https://github.com/influxdata/influxdb_iox/blob/e82214e/iox_query/src/exec/query_tracing.rs#L111-L211

**Comment by crepererum** [comment]
I think SOME information should be send via HTTP headers, esp. data that we have before we start the result streaming:

- queue time / "semaphore acquire duration"
- information about the plan
- ...

Data that is collected during execution and result streaming shall be send via HTTP trailers:

- row counts
- execution time
- ...

I have a rough idea on how to do both of them. @mhilton do you mind if I steal this ticket from you?

**Comment by mhilton** [comment]
I certainly don't mind you taking the ticket off me. I don't have time to do it soon and would like it done.

**Comment by mhilton** [comment]
Just as a point on the implementation, shouldn't this be implemented in terms of gRPC metadata rather than specifically HTTP Headers / Trailers (even if that's what gRPC does with them)?

**Comment by crepererum** [comment]
From reading tonic and the gRPC go impl, I think "gRPC metadata" is the common term that refers both to HTTP/2 headers and trailers and we can use both (depending on what/when we wanna transmit information). It's just that the trailers aren't wired very well into tonic, but this shouldn't stop us (I have a prototype using a tower middleware that ends up in a nice API).

**Comment by crepererum** [comment]
Since every feature needs a test, I figured that the flight client probably should be able to read headers/trailers. This will be possible after https://github.com/apache/arrow-rs/pull/4727 is available in IOx, until then I think this ticket here should be parked.

**Comment by alamb** [comment]
> https://github.com/apache/arrow-rs/pull/4727

This should be in `arrow-flight 47.0.0` eta in IOx is around 2 weeks

**Comment by crepererum** [comment]
I plan to work on that next week. I have a local branch that implements a rough version of it, but needs tests and a proper code clean-ups.
